xayn-ai/src/ranker/context.rs

```diff
@@ -1,119 +1,231 @@
-        };
-
-        // In the `assert_approx_eq!`s below, the result expectations are
-        // expressions instead of constants to make it easier to understand how
-        // the values come into existence. The format is (frac_pos + frac_neg) / 2.
-
-        let cxt = calc.calculate_score(0., calc.neg_max);
-        assert_approx_eq!(f32, cxt, (1. + 1.) / 2.);
-
-        let cxt = calc.calculate_score(1., calc.neg_max);
-        assert_approx_eq!(f32, cxt, ((4. / 5.) + 1.) / 2.);
+    use crate::{
+        coi::{create_neg_cois, create_pos_cois},
+        utils::SECONDS_PER_DAY,
+    };
 
-        let cxt = calc.calculate_score(calc.pos_avg, calc.neg_max);
-        assert_approx_eq!(f32, cxt, (1. / 2. + 1.) / 2.);
+    use super::*;
 
-        let cxt = calc.calculate_score(8., 7.);
-        assert_approx_eq!(f32, cxt, ((1. / 3.) + (1. / 2.)) / 2.);
+    #[test]
+    fn test_has_enough_cois() {
+        let user_interests = UserInterests::default();
+
+        assert!(has_enough_cois(&user_interests, 0, 0).is_ok());
+        assert!(matches!(
+            has_enough_cois(&user_interests, 1, 0).unwrap_err(),
+            Error::NotEnoughPositiveCois {
+                has: 0,
+                expected: 1
+            }
+        ));
+        assert!(matches!(
+            has_enough_cois(&user_interests, 0, 1).unwrap_err(),
+            Error::NotEnoughNegativeCois {
+                has: 0,
+                expected: 1
+            }
+        ));
     }
 
-    #[allow(clippy::eq_op)]
     #[test]
-    fn test_calculate_neg_max_f32_max() {
-        // when calculating the negative distance in the `CoiSystem`,
-        // we assign `f32::MAX` if we don't have negative cois
-        let calc = Context {
-            pos_avg: 4.,
-            neg_max: f32::MAX,
-        };
-
-        let ctx = calc.calculate_score(0., calc.neg_max);
-        assert_approx_eq!(f32, ctx, (1. + 1.) / 2.);
+    fn test_compute_score_for_embedding() {
+        let embedding = arr1(&[0., 0., 0.]).into();
+
+        let epoch = SystemTime::UNIX_EPOCH;
+        let now = epoch + Duration::from_secs_f32(2. * SECONDS_PER_DAY);
+
+        let mut positive = create_pos_cois(&[[1., 2., 3.], [4., 5., 6.]]);
+        positive[0].stats.last_view -= Duration::from_secs_f32(0.5 * SECONDS_PER_DAY);
+        positive[1].stats.last_view -= Duration::from_secs_f32(1.5 * SECONDS_PER_DAY);
+
+        let mut negative = create_neg_cois(&[[100., 0., 0.]]);
+        negative[0].last_view = epoch;
+        let user_interests = UserInterests { positive, negative };
+
+        let mut relevances = RelevanceMap::default();
+        let horizon = Duration::from_secs_f32(2. * SECONDS_PER_DAY);
+
+        let score = compute_score_for_embedding(
+            &embedding,
+            &user_interests,
+            &mut relevances,
+            1,
+            horizon,
+            now,
+        )
+        .unwrap();
+        // 1.1185127 * 0.99999934 + 0.73094887 - 0 * 100 = 12
```

Comment:
what does the `= 12` refer to?

Reply:
That was the result of the old calculation. I forgot to update it

Reply:
That was the result of the old calculation. I forgot to update it

Reply:
done